### PR TITLE
Rework of `dtype`s

### PR DESCRIPTION
**Pull request by tudoroancea**
In this PR:
- we add notion of dtype precision (32 or 64 bits) with the enum `DTypePrecision`
- we define *compatible backends* as backends with the same type (i.e. the implementation) and dtype precision (but that may differ in realness)
- we enforce compatibility of backends between functions (cost and manually specified derivatives) and manifolds in a problem
- we enforce compatibility of backends of manifolds in a product, and now add a backend for the product itself. It is directly determined from the backends of the underlying manifolds and was (arbitrarily) chosen convention to be the real version compatible with the others (which should not impact anything).
- we replace the old `Manifold.backend.setter` by a method `Manifold.set_compatible_backend(other_backend)` that is a safe setter enforcing backend compatibility
- we allow now specifying functions (cost and its derivatives) as pure python functions and then convert them to `Function` inside the `Problem` based on the backend of the `Manifold`. We raise an exception if neither the functions nor the manifold have a backend.

> [!NOTE]
> This PR (#284) is stacked with #285 and #286 (and is at the bottom of the stack). This means that the branch of #285 is based on the one of #284 and the branch of #286 on the branch of #285. They each treat different issues in the codebase and were split to facilitate reviewing. However, there are still unit tests failing in the first ones, and they are only totally fixed in the last. Thanks for taking that into account during the review.

### CODE REVIEW

**Comment by antoinecollas** [comment]
The tests are not run anymore automatically in this PR. Do you have an idea why?

File: src/pymanopt/backends/numpy_backend.py [hunk 1]
```diff
@@ -35,6 +35,11 @@ def __init__(self, dtype: type = np.float64):
             or dtype == np.complex128
         ), f"dtype {dtype} is not supported"
         self._dtype = dtype
+        self._dtype_precision = (
+            DTypePrecision.SINGLE
+            if (dtype == np.float32 or dtype == np.complex64)
+            else DTypePrecision.DOUBLE
+        )
```
Comment (antoinecollas) [inline src/pymanopt/backends/numpy_backend.py]: If you dynamically change `self._dtype` then `self._dtype_precision` is not changed accordingly. I think it's better to define a getter for `self._dtype_precision` to avoid this.

File: src/pymanopt/backends/pytorch_backend.py [hunk 2]
```diff
@@ -47,6 +47,11 @@ def __init__(self, dtype=torch.float64):
             torch.complex128,
         }
         self._dtype = dtype
+        self._dtype_precision = (
+            DTypePrecision.SINGLE
+            if dtype in {torch.float32, torch.complex64}
+            else DTypePrecision.DOUBLE
+        )
```
Comment (antoinecollas) [inline src/pymanopt/backends/pytorch_backend.py]: same problem

File: src/pymanopt/backends/tensorflow_backend.py [hunk 3]
```diff
@@ -48,7 +48,18 @@ class TensorflowBackend(Backend):
...
+            tf.float64,
+            tf.complex64,
+            tf.complex128,
+        }, f"dtype {dtype} is not supported"
         self._dtype = dtype
+        self._dtype_precision = (
+            DTypePrecision.SINGLE
+            if dtype in {tf.float32, tf.complex64}
+            else DTypePrecision.DOUBLE
+        )
```
Comment (antoinecollas) [inline src/pymanopt/backends/tensorflow_backend.py]: same problem

File: src/pymanopt/backends/backend.py [hunk 4]
```diff
@@ -30,15 +31,23 @@ def inner(self, *args, **kwargs):
...
     """Abstract base class defining the interface for autodiff backends."""
 
     ##########################################################################
     # Common attributes, properties and methods
     ##########################################################################
     _dtype: type
+    _dtype_precision: DTypePrecision
 
-    @runtime_checkable
+    @runtime_checkable  # to be able to check isinstance(x, bk.array_t)
```
Comment (antoinecollas) [inline src/pymanopt/backends/backend.py:49]: why did you add this comment?

File: src/pymanopt/optimizers/nelder_mead.py [hunk 5]
```diff
@@ -12,14 +12,12 @@
...
 
-    @pymanopt.function.numpy(manifold)
     def objective(*y):
         if manifold.num_values == 1:
             (y,) = y
         return sum(manifold.dist(y, point) ** 2 for point in points) / 2
 
-    @pymanopt.function.numpy(manifold)
-    def gradient(*y):
+    def riemannian_gradient(*y):
```
Comment (antoinecollas) [inline src/pymanopt/optimizers/nelder_mead.py:20]: why did you change the name of the function?

File: src/pymanopt/manifolds/sphere.py [hunk 6]
```diff
@@ -202,12 +202,12 @@ def __init__(
...
-        super().backend = backend
-        self._matrix = backend.array(self._matrix)
-        q, _ = backend.linalg_qr(self._matrix)
-        self._subspace_projector = q @ self.backend.transpose(q)
+    def set_compatible_backend(self, other_backend: Backend):
+        super().set_compatible_backend(other_backend)
+        bk = self.backend
+        self._matrix = bk.array(self._matrix)
+        q, _ = bk.linalg_qr(self._matrix)
+        self._subspace_projector = q @ bk.transpose(q)
```
Comment (antoinecollas) [inline src/pymanopt/manifolds/sphere.py]: it seems to be redundant with the what's inside the __init__. Is it possible to factorize?

File: src/pymanopt/manifolds/euclidean.py [hunk 7]
```diff
@@ -107,11 +107,6 @@ def __init__(
         dimension = math.prod(shape)
         super().__init__(name, dimension, *shape, backend=backend)
 
-        @RiemannianSubmanifold.backend.setter
-        def _(self, backend: Backend):
-            assert backend.is_dtype_real()
-            super().backend = backend
-
 
 class ComplexEuclidean(_Euclidean):
```
Comment (antoinecollas) [inline src/pymanopt/manifolds/euclidean.py:111]: With what you propose, maybe it's time to get rid of these `ComplexXXXX` manifolds?

**Review by antoinecollas — COMMENTED** [review]
Here is a quick review

File: src/pymanopt/optimizers/nelder_mead.py (hunk 5 above)
Comment (tudoroancea) [reply src/pymanopt/optimizers/nelder_mead.py:20]: Simply because I saw below that the function was used as a `riemannian_gradient` when constructing the `Problem`. But maybe I misinterpreted something

File: src/pymanopt/backends/backend.py (hunk 4 above)
Comment (tudoroancea) [reply src/pymanopt/backends/backend.py:49]: I felt I needed to explain what 'checkable' meant in this context so that devs would not need to read the python docu. Happy to remove it if you judge it unnecessary

File: src/pymanopt/manifolds/euclidean.py (hunk 7 above)
Comment (tudoroancea) [reply src/pymanopt/manifolds/euclidean.py:111]: Here my thoughts on the matter:
- The manifold dimension still changes between the real and complex case. I think we could work around this by making the dimension an `@property`.
- Currently, each manifold class has an `IS_COMPLEX` attribute indicating its realness. It is used in the `set_compatible_backend()` to choose what version of the backend (real or complex) to use. Note there are manifolds that can have both real and complex dtypes (e.g. `Euclidean`), but there also manifolds that can only be one (e.g. `SymmetricPositiveDefinite` can only be real and ` SpecialHermitianPositiveDefinite` can only be complex). We should also be able to deal with this but it would require replacing `IS_COMPLEX` with (for example) an attribute listing the possible realness of the manifold.
- Nevertheless, removing these classes directly would be a huge breaking change, and the effort required to mitigate it (e.g. by declaring a bunch of deprecated wrappers) would be considerable.

Overall, this would be a nice product of the work around the numerical backends. However, it is still independent of our current objectives and I think we can postpone until after the [`develop_backends`](https://github.com/pymanopt/pymanopt/tree/develop_backends) branch is merged.

File: src/pymanopt/function.py [hunk 8]
```diff
@@ -70,12 +77,17 @@ def inner(cost: Callable[..., Any]) -> Function:
...
                 import_module(
                     f"pymanopt.backends.{module}",
                 ),
                 backend_class,
-            )()
+            )
+            backend = (
+                backend_type(dtype=dtype)
+                if dtype is not None
+                else backend_type()
```
Comment (tudoroancea) [inline src/pymanopt/function.py:91]: comment that using the real version doesn't change anything for a function because the autodiff methods are the same

File: src/pymanopt/manifolds/product.py [hunk 9]
```diff
@@ -73,19 +70,11 @@ def __init__(
...
             point_layout=point_layout,
-            backend=manifolds[0].backend,
-        )
-
-    @Manifold.backend.setter
-    def backend(self, backend: Backend):
-        warnings.warn(
-            "Setting backend of Product manifold is not supported. "
-            "One should directly set the backend of each underlying manifold."
+            backend=manifolds[0].backend.to_real_backend(),
```
Comment (tudoroancea) [inline src/pymanopt/manifolds/product.py:80]: add a comment to say real doesn't matter here (only used to instantiate the backend of a function)

File: src/pymanopt/optimizers/nelder_mead.py [hunk 10]
```diff
@@ -12,14 +12,12 @@
 def compute_centroid(manifold, points):
     """Compute the centroid of `points` on the `manifold` as Karcher mean."""
```
Comment (tudoroancea) [inline src/pymanopt/optimizers/nelder_mead.py]: Suggestion of @NicolasBoumal:
```suggestion
    """Compute the centroid of `points` on the `manifold` as intrinsic mean."""
```

**Comment by NicolasBoumal** [comment]
Merging this now (follow up PRs #285 and #286 will take care of the remaining tests that fail).
